File: internal/cli/project/update.go
package project

import (
	"context"

	"github.com/spf13/cobra"

	"github.com/shhac/lin/internal/linear"
	"github.com/shhac/lin/internal/output"
	"github.com/shhac/lin/internal/resolvers"
)

func registerUpdate(parent *cobra.Command) {
	update := &cobra.Command{
		Use:   "update",
		Short: "Update project fields",
	}
	parent.AddCommand(update)

	registerUpdateTitle(update)
	registerUpdateStatus(update)
	registerUpdateLead(update)
	registerUpdatePriority(update)

	registerSimpleProjectUpdate(update, "description <id> <description>", "Update project description",
		func(v string) linear.ProjectUpdateInput { return linear.ProjectUpdateInput{Description: &v} })
	registerSimpleProjectUpdate(update, "content <id> <content>", "Update project content (markdown body)",
		func(v string) linear.ProjectUpdateInput { return linear.ProjectUpdateInput{Content: &v} })
	registerSimpleProjectUpdate(update, "start-date <id> <date>", "Update project start date",
		func(v string) linear.ProjectUpdateInput { return linear.ProjectUpdateInput{StartDate: &v} })
	registerSimpleProjectUpdate(update, "target-date <id> <date>", "Update project target date",
		func(v string) linear.ProjectUpdateInput { return linear.ProjectUpdateInput{TargetDate: &v} })
	registerSimpleProjectUpdate(update, "icon <id> <icon>", "Update project icon",
		func(v string) linear.ProjectUpdateInput { return linear.ProjectUpdateInput{Icon: &v} })
	registerSimpleProjectUpdate(update, "color <id> <color>", "Update project color",
		func(v string) linear.ProjectUpdateInput { return linear.ProjectUpdateInput{Color: &v} })

	output.HandleUnknownCommand(update, "Run `lin project usage` for available update subcommands")
}

func registerSimpleProjectUpdate(parent *cobra.Command, use, short string, buildInput func(string) linear.ProjectUpdateInput) {
	parent.AddCommand(&cobra.Command{
		Use:   use,
		Short: short,
		Args:  cobra.ExactArgs(2),
		Run: func(cmd *cobra.Command, args []string) {
			client := linear.GetClient()
			ctx := context.Background()

			resolved, err := resolvers.ResolveProject(client, args[0])
			if err != nil {
				output.PrintError(err.Error())
			}

			resp, err := linear.ProjectUpdate(ctx, client, resolved.ID, buildInput(args[1]))
			if err != nil {
				output.HandleGraphQLError(err)
			}

			output.PrintJSON(map[string]any{"updated": resp.ProjectUpdate.Success})
		},
	})
}

File: bun/src/cli/document/update.ts
import type { Command } from "commander";
import { getClient } from "../../lib/client.ts";
import { printError, printJson } from "../../lib/output.ts";
import { resolveDocument, resolveProject } from "../../lib/resolvers.ts";

export function registerUpdate(document: Command): void {
  const update = document.command("update").description("Update document fields");

  update
    .command("title")
    .description("Update document title")
    .argument("<id>", "Document ID or slug ID")
    .argument("<new-title>", "New title")
    .action(async (id: string, newTitle: string) => {
      try {
        const client = getClient();
        const doc = await resolveDocument(client, id);
        const payload = await client.updateDocument(doc.id, { title: newTitle });
        printJson({ updated: payload.success });
      } catch (err) {
        printError(err instanceof Error ? err.message : "Update failed");
      }
    });

  update
    .command("content")
    .description("Update document content")
    .argument("<id>", "Document ID or slug ID")
    .argument("<content>", "New content (markdown)")
    .action(async (id: string, content: string) => {
      try {
        const client = getClient();
        const doc = await resolveDocument(client, id);
        const payload = await client.updateDocument(doc.id, { content });
        printJson({ updated: payload.success });
      } catch (err) {
        printError(err instanceof Error ? err.message : "Update failed");
      }
    });

  update
    .command("project")
    .description("Move document to project")
    .argument("<id>", "Document ID or slug ID")
    .argument("<project>", "Project ID, slug, or name")
    .action(async (id: string, project: string) => {
      try {
        const client = getClient();
        const [doc, resolved] = await Promise.all([
          resolveDocument(client, id),
          resolveProject(client, project),
        ]);
        const payload = await client.updateDocument(doc.id, { projectId: resolved.id });
        printJson({ updated: payload.success });
      } catch (err) {
        printError(err instanceof Error ? err.message : "Update failed");
      }
    });

  update
    .command("icon")
    .description("Update document icon")
    .argument("<id>", "Document ID or slug ID")
    .argument("<icon>", "Icon (emoji)")
    .action(async (id: string, icon: string) => {
      try {
        const client = getClient();
        const doc = await resolveDocument(client, id);
        const payload = await client.updateDocument(doc.id, { icon });
        printJson({ updated: payload.success });
      } catch (err) {
        printError(err instanceof Error ? err.message : "Update failed");
      }
    });

  update
    .command("color")
    .description("Update document color")
    .argument("<id>", "Document ID or slug ID")
    .argument("<color>", "Color (hex, e.g. #5e6ad2)")
    .action(async (id: string, color: string) => {
      try {
        const client = getClient();
        const doc = await resolveDocument(client, id);
        const payload = await client.updateDocument(doc.id, { color });
        printJson({ updated: payload.success });
      } catch (err) {
        printError(err instanceof Error ? err.message : "Update failed");
      }
    });
}
